### PR TITLE
Add `DNSTLDResolver` and `DNSTXTResolver`

### PR DESCRIPTION
**Pull request by adraffy**
* added [DNSTLDResolver.sol](https://github.com/ensdomains/namechain/blob/feat/bet-379-dns-resolver/contracts/src/L1/dns/DNSTLDResolver.sol) and [DNSTLDResolver.test.ts](https://github.com/ensdomains/namechain/blob/feat/bet-379-dns-resolver/contracts/test/dns/DNSTLDResolver.test.ts) and [DNSTLDResolver.t.sol](https://github.com/ensdomains/namechain/blob/feat/bet-379-dns-resolver/contracts/test/dns/DNSTLDResolver.t.sol)
    * supports `resolve(multicall)`
    * supports chunked TXT
    * added `setOracleGateways()`
    * added `setBatchGateways()`

---

* added [DNSTXTResolver.sol](https://github.com/ensdomains/namechain/blob/feat/bet-379-dns-resolver/contracts/src/L1/dns/DNSTXTResolver.sol) (tests are part of [DNSTLDResolver.test.ts](https://github.com/ensdomains/namechain/blob/feat/bet-379-dns-resolver/contracts/test/dns/DNSTLDResolver.test.ts#L256-L453))
    * supports `resolve(multicall)`
    * supports default address
    * added `c=` for `contenthash()`
    * added `xy=` for `pubkey()`
    * added `text("eth.ens.dnssec-context")` passthru
* added [DNSTXTParser.sol](https://github.com/ensdomains/namechain/blob/feat/bet-379-dns-resolver/contracts/src/L1/dns/DNSTXTParser.sol) and [DNSTXTParser.t.sol](https://github.com/ensdomains/namechain/blob/feat/bet-379-dns-resolver/contracts/test/dns/DNSTXTParser.t.sol)
    * backwards compatible with previous parser
    * [`test_find_quotedWithoutGap()`](https://github.com/ensdomains/namechain/blob/448cae21329f05058bf8f0d9d46f004c1690d909/contracts/test/dns/DNSTXTScanner.t.sol#L47) is a discovered edge case

---

* ~~temporary `RegistryUtilsV1.sol`~~ &rarr; https://github.com/ensdomains/ens-contracts/pull/462
* added [RegistryUtils.sol](https://github.com/ensdomains/namechain/blob/feat/bet-379-dns-resolver/contracts/src/universalResolver/RegistryUtils.sol)
     * added `findExactRegistry()` — was `UR.getRegistry()`
     * added `findParentRegistry()` — was `UR.getParentRegistry()`
     * added `readLabel()` — was `NameUtils.readLabel()`
     * ~~temporary `nextLabel()`~~ &rarr; https://github.com/ensdomains/ens-contracts/pull/459
* changed `UniversalResolverTraversal.t.sol` to [RegistryUtils.t.sol](https://github.com/ensdomains/namechain/blob/feat/bet-379-dns-resolver/contracts/test/universalResolver/RegistryUtils.t.sol) and added more tests
* changed `Ens.test.ts` to [deployFixtureV2.test.ts](https://github.com/ensdomains/namechain/blob/feat/bet-379-dns-resolver/contracts/test/deployV2Fixture.test.ts) and added more tests
* changed `AUR` and `DNSTLDResolver` to use the finders
     * future PR &rarr; change `ETHFallbackResolver` to use finders
     * future PR &rarr; put forward resolver into abstract class

---

* synced AUR from `ens-contracts/`
* bumped `ens-contracts/` and adjusted imports
     * use features
     * use `IPriceOracle`
* added [supportsFeatures.ts](https://github.com/ensdomains/namechain/blob/feat/bet-379-dns-resolver/contracts/test/utils/supportsFeatures.ts)
* switched to `mcopy` and set semver accordingly
* moved some utilities around
* added `dns/` subdirectories
* ~~added `HardhatIgnoreWarningsPlugin` and config~~ &rarr; https://github.com/ensdomains/namechain/pull/80

### CODE REVIEW

File: contracts/src/L1/dns/DNSRegistryResolver.sol [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    function _parseContext(
+        bytes calldata context
+    ) internal pure returns (address registry, bytes32 nodeSuffix) {
+        require(
+            context.length >= 43 &&
+                context[0] == "0" &&
+                context[1] == "x" &&
+                context[42] == " ",
+            "expected <address> <suffix>"
```
Comment (Arachnid) [inline contracts/src/L1/dns/DNSRegistryResolver.sol:64]: We should use error objects rather than strings.

File: contracts/src/L1/dns/DNSRegistryResolver.sol
Comment (Arachnid) [inline contracts/src/L1/dns/DNSRegistryResolver.sol:1]: Can you explain when you'd use this instead of the alias resolver? Perhaps this can also be moved to its own PR?

**Review by Arachnid — CHANGES_REQUESTED** [review]
I've left a few comments, but this will be much easier to review if you're able to split this into smaller PRs we can review sequentially, starting with the core components.

File: contracts/src/common/IRegistryResolver.sol
[the remark this answers is not in the document]
Comment (adraffy) [reply contracts/src/common/IRegistryResolver.sol:1]: No, it's a property of ETHFallbackResolver or any other "ENS Registry" gateway

File: contracts/src/L1/ETHFallbackResolver.sol
[the remark this answers is not in the document]
Comment (adraffy) [reply contracts/src/L1/ETHFallbackResolver.sol:1]: Yeah, I can probably split out `ETHFallbackResolver`, `IRegistryResolver`, and `DNSRegistryResolver`. 

`DNSRegistryResolver` just seemed like a natural addition to the gasless arsenal.

File: contracts/src/common/IRegistryResolver.sol [hunk 2]
```diff
@@ -0,0 +1,18 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.13;
+
+interface IRegistryResolver {
+    /// @notice Resolve `name` staring at `parentRegistry` using the labels after `nodeSuffix`.
+    /// @notice Caller should enable EIP-3668.
+    /// @param parentRegistry The parent registry contract.
+    /// @param nodeSuffix The node corresponding to the parent registry contract.
```
[the remark this answers is not in the document]
Comment (adraffy) [reply contracts/src/common/IRegistryResolver.sol]: Yeah, that's the direction I was moving towards.  I just refactored out the parsing component into `NameMatcher.sol` which means this signature could be reduced to `resolve(name, offset, registry, data)` where it uses `name` up to `offset`.

Anyway, this was moved to a follow-up PR.

File: contracts/src/common/IRegistryResolver.sol [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    /// @param name The name to resolve.
+    /// @param data The calldata.
```
[the remark this answers is not in the document]
Comment (adraffy) [reply contracts/src/common/IRegistryResolver.sol]: Improved in new PR.

File: contracts/src/L1/dns/DNSRegistryResolver.sol (hunk 1 above)
Comment (adraffy) [reply contracts/src/L1/dns/DNSRegistryResolver.sol]: I'll bring this up again when the PR is reposted.  I was planning on replacing this with a typed error, but then I had the idea that possibly these errors should be strings since they're user-facing.  I only left it like this since `DNSAliasResolver` and `DNSRegistryResolver` were new creations.

File: contracts/src/L1/dns/DNSRegistryResolver.sol
Comment (adraffy) [reply contracts/src/L1/dns/DNSRegistryResolver.sol:1]: The general idea is that you can map `*.nick.com` to any registry namespace on Namechain, including detached ones.

The remapping rule is powerful, such that `a.b.c.nick.com` with `ENS1 <DNSRegistryResolver> <parentRegistry> nick.com` will effectively be: `parentRegistry.getRegistry("c").getRegistry("b").getResolver("a")`

File: contracts/src/common/IRegistryResolver.sol [hunk 4]
```diff
@@ -0,0 +1,18 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.13;
+
+interface IRegistryResolver {
+    /// @notice Resolve `name` staring at `parentRegistry` using the labels after `nodeSuffix`.
+    /// @notice Caller should enable EIP-3668.
```
[the remark this answers is not in the document]
Comment (adraffy) [reply contracts/src/common/IRegistryResolver.sol]: At the moment, the only crosschain registry is on Namechain, so `ETHFallbackResolver` is the only `IRegistryResolver`, but other chains could work, and you're correct they all require EIP-3668.

However, technically, with the help of an `IRegistryResolver` on mainnet, you could use the `DNSRegistryResolver` to link to a V2 registry that isn't reachable from root and that would not require EIP-3668.  For any registry that's reachable, the `DNSAliasResolver` would work since it's just a name transformation.

File: contracts/src/L1/ETHFallbackResolver.sol [hunk 5]
```diff
@@ -101,90 +97,129 @@ contract ETHFallbackResolver is
...
-    /// @return calls The individual calls.
-    function _parseCalls(
+    /// @dev Determine if labelhash is actively registered on V1.
+    /// @param id The labelhash of the "eth" 2LD.
+    /// @return True if the registration is active.
+    function _isActiveRegistrationV1(uint256 id) internal view returns (bool) {
+        return
+            !ethRegistrarV1.available(id) &&
+            ethRegistrarV1.ownerOf(id) != burnAddressV1;
+    }
```
[the remark this answers is not in the document]
Comment (adraffy) [reply contracts/src/L1/ETHFallbackResolver.sol]: We discussed this in [DedicatedResolver PR](https://github.com/ensdomains/namechain/pull/58#discussion_r2144416105).  It should be simple to change this to whatever we need.  I can put up a PR or two that experiments with those ideas + some tests.

File: contracts/src/L1/dns/DNSTLDResolver.sol [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+    function _parseResolver(
+        bytes memory v
+    ) internal view returns (address resolver) {
+        if (v.length == 42 && v[0] == "0" && v[1] == "x") {
+            (address addr, bool valid) = HexUtils.hexToAddress(v, 2, 42);
+            if (valid) {
+                return addr;
+            }
+        }
+        (resolver, , ) = universalResolverV2.findResolver(
```
Comment (Arachnid) [inline contracts/src/L1/dns/DNSTLDResolver.sol:354]: Likewise for finding the v2 resolver - can we make this a library function?

Comment (adraffy) [reply contracts/src/L1/dns/DNSTLDResolver.sol:354]: Oh, instead of using the instance?  Yeah that's likely a better design.

Comment (adraffy) [reply contracts/src/L1/dns/DNSTLDResolver.sol:354]: Oh, the issue is that you need `rootRegistry` for V2 and the `registry` for V1.  So even if `findResolver()` had a library, you'd still have to supply those addresses unless we embed them in the helper.

File: contracts/src/L1/dns/DNSTLDResolver.sol [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+        (address resolver, , uint256 offset) = universalResolverV1.findResolver(
+            name
+        );
+        if (
+            resolver != address(0) &&
+            (offset == 0 ||
+                ERC165Checker.supportsERC165InterfaceUnchecked(
+                    resolver,
+                    type(IExtendedResolver).interfaceId
+                ))
```
[the remark this answers is not in the document]
Comment (adraffy) [reply contracts/src/L1/dns/DNSTLDResolver.sol:127]: Oh right, good catch.  Changed to an address check.

**Comment by adraffy** [comment]
There are no `AUR` changes, it's just syncing the code from `ens-contracts/`.   If we merge [this PR](https://github.com/ensdomains/namechain/pull/74) we might be able to use imported `AUR` instead (with some fancy remapping logic due to the OZ 4 vs 5 issue.)

File: contracts/src/L1/dns/DNSTLDResolver.sol [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+        ) {
+            _callResolver(resolver, name, data, false, "");
+        }
```
[the remark this answers is not in the document]
Comment (adraffy) [reply contracts/src/L1/dns/DNSTLDResolver.sol]: There is also UR V1 dependency in `ETHFallbackResolver`, should I remove that too?

[the remark this answers is not in the document]
Comment (adraffy) [reply contracts/src/L1/dns/DNSTLDResolver.sol]: I guess instead of referencing a non-proxy UR v2, there could be an contract base, say `ResolverCaller`, which implements the `_callResolver()` w/o `IExtendedDNSResolver` (basically the core part of AUR), and then the AUR, ETHFallbackResolver, and any other UR-user could just inherit that contract instead.

File: contracts/src/L1/dns/DNSTLDResolver.sol [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+contract DNSTLDResolver is
+    IFeatureSupporter,
+    IExtendedResolver,
+    CCIPBatcher,
+    Ownable,
+    ERC165
+{
+    IUniversalResolverStub public immutable universalResolverV1;
+    IUniversalResolverStub public immutable universalResolverV2;
+    DNSSEC public immutable oracleVerifier;
```
[the remark this answers is not in the document]
Comment (adraffy) [reply contracts/src/L1/dns/DNSTLDResolver.sol]: Changed

File: contracts/src/L1/dns/DNSTXTResolver.sol [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+import {IAddressResolver} from "@ens/contracts/resolvers/profiles/IAddressResolver.sol";
+import {IHasAddressResolver} from "@ens/contracts/resolvers/profiles/IHasAddressResolver.sol";
+import {ITextResolver} from "@ens/contracts/resolvers/profiles/ITextResolver.sol";
+import {IContentHashResolver} from "@ens/contracts/resolvers/profiles/IContentHashResolver.sol";
+import {IPubkeyResolver} from "@ens/contracts/resolvers/profiles/IPubkeyResolver.sol";
+
+/// @dev The text key to access "context" from `ENS1 <resolver> <context>`.
+string constant TEXT_DNSSEC_CONTEXT = "eth.ens.dnssec-context";
+
+contract DNSTXTResolver is ERC165, IFeatureSupporter, IExtendedDNSResolver {
```
[the remark this answers is not in the document]
Comment (adraffy) [reply contracts/src/L1/dns/DNSTXTResolver.sol:49]: https://github.com/ensdomains/namechain/blob/feat/bet-379-dns-resolver/contracts/src/L1/dns/DNSTXTResolver.sol#L27-L48

File: contracts/src/L1/dns/DNSTLDResolver.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (adraffy) [reply contracts/src/L1/dns/DNSTLDResolver.sol]: https://github.com/ensdomains/namechain/blob/feat/bet-379-dns-resolver/contracts/src/L1/dns/DNSTLDResolver.sol#L130-L137

File: contracts/src/universalResolver/AbstractUniversalResolver.sol [hunk 11]
```diff
@@ -78,148 +94,164 @@ abstract contract AbstractUniversalResolver is
...
+        view
+        returns (
+            string memory primary,
+            address resolver,
+            address reverseResolver
+        )
+    {
+        primary;
+        resolver;
+        reverseResolver;
```
Comment (TateB) [inline contracts/src/universalResolver/AbstractUniversalResolver.sol:206]: what is this for?

File: contracts/test/utils/supportsFeatures.ts [hunk 12]
```diff
@@ -0,0 +1,44 @@
+import { toFunctionSelector, type Hex } from "viem";
+import { describe, it, expect } from "vitest";
+import { FEATURES } from "../../lib/ens-contracts/test/utils/features.js";
+
+type FeatureKeys<T> = {
+  [K in keyof T]?: readonly (keyof T[K])[];
+};
+
+export function shouldSupportsFeatures({
```
Comment (TateB) [inline contracts/test/utils/supportsFeatures.ts:9]: shouldSupportFeatures

**Review by TateB — CHANGES_REQUESTED** [review]

File: contracts/src/universalResolver/RegistryUtils.sol [hunk 13]
```diff
@@ -0,0 +1,132 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.24;
+
+import {IRegistry} from "../common/IRegistry.sol";
+import {NameCoder} from "@ens/contracts/utils/NameCoder.sol";
+
+library RegistryUtils {
+    /// @notice Find the resolver address for `name[offset:]`.
```
Comment (Arachnid) [inline contracts/src/universalResolver/RegistryUtils.sol:8]:
```suggestion
    /// @dev Find the resolver address for `name[offset:]`.
```
I think this is the correct tag for these functions' short descriptions.

File: contracts/src/universalResolver/RegistryUtils.sol [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+    /// @param rootRegistry The root ENS registry.
+    /// @param name The DNS-encoded name to search.
+    /// @param offset The offset into `name` to begin the search.
+    /// @return registry The nearest registry for `name` (not the parent registry).
```
Comment (Arachnid) [inline contracts/src/universalResolver/RegistryUtils.sol:12]: Why 'not the parent registry'?

File: contracts/src/universalResolver/RegistryUtils.sol [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+        assembly {
+            mcopy(add(label, 32), add(add(name, 33), offset), size)
+        }
+    }
+
+    // ********************************************************************************
+    // I'd like to move the following code here:
+    // https://github.com/ensdomains/ens-contracts/pull/459
+    // https://github.com/ensdomains/ens-contracts/blob/fix/namecoder-iter/contracts/utils/NameCoder.sol#L42
+    // ********************************************************************************
```
Comment (Arachnid) [inline contracts/src/universalResolver/RegistryUtils.sol:111]: If this is something you can't do right now, label it as a TODO with a specific condition when it can happen.

File: contracts/src/L1/dns/DNSTXTScanner.sol [hunk 16]
```diff
@@ -0,0 +1,162 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.13;
+
+import {BytesUtils} from "@ens/contracts/utils/BytesUtils.sol";
+
+/// @title DNSTXTScanner
```
[the remark this answers is not in the document]
Comment (adraffy) [reply contracts/src/L1/dns/DNSTXTScanner.sol:6]: Changed

File: contracts/src/L1/dns/DNSTLDResolver.sol [hunk 17]
```diff
@@ -0,0 +1,412 @@
...
+    /// @dev Parse the TXT record into resolver and context.
+    ///      Format: "ENS1 <name-or-address> <context?>".
+    /// @param txt The TXT data.
+    /// @return resolver The resolver address or null if wrong format or name didn't resolve.
+    /// @return context The optional context data.
+    function _parseTXT(
+        bytes memory txt
+    ) internal view returns (address resolver, bytes memory context) {
+        uint256 n = TXT_PREFIX.length;
+        if (txt.length >= n && BytesUtils.equals(txt, 0, TXT_PREFIX, 0, n)) {
```
[the remark this answers is not in the document]
Comment (adraffy) [reply contracts/src/L1/dns/DNSTLDResolver.sol:337]: I contemplated changing `BytesUtils.equals()` when I made a pass over it (but I didn't want to change any behavior, just add more revert information.)  It's defined as two sliced `BytesUtils.keccak()`'s eg. `keccak() == keccak()`, which will revert if either of them is out of bounds.

This ensures it doesn't revert.

**Review by TateB — APPROVED** [review]
